### PR TITLE
layer access stats for layer eviction

### PR DESCRIPTION
**Pull request by problame**
```
    maintain access stats for historic layers
    
    This patch adds basic access statistics for historic layers
    and exposes them in the management API's `LayerMapInfo`.
    
    We record the accesses in the `{Delta,Image}Layer::load()` function
    because it's the common path of
    * page_service (`Timline::get_reconstruct_data()`)
    * Compaction (`PersistentLayer::iter()` and `PersistentLayer::key_iter()`)
    
    There is a killswitch in pageserver.toml to disable statistics gathering
    to support before/after benchmarks. With the killswitch enabled, the
    overhead of this feature is the function call and the atomic load.
    
    Preceding PRs:
    - #3504 add RequestContext plumbing for layer access stats
    
    This is PR https://github.com/neondatabase/neon/pull/3496
```

Preceding patches:
- https://github.com/neondatabase/neon/pull/3502
- https://github.com/neondatabase/neon/pull/3510
- https://github.com/neondatabase/neon/pull/3504

### CODE REVIEW

File: pageserver/src/tenant/storage_layer.rs [hunk 1]
```diff
@@ -81,6 +86,54 @@ pub enum ValueReconstructResult {
     Missing,
 }
 
+#[derive(Default)]
+pub struct LayerAccessStats(RwLock<LayerAccessStatsInner>);
+
+#[derive(Default)]
+struct LayerAccessStatsInner {
+    last_access: Option<SystemTime>,
+    first_access: Option<SystemTime>,
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer.rs:95]: With this mutex construction we can just use Instants everywhere.

File: pageserver/src/tenant/storage_layer/delta_layer.rs [hunk 2]
```diff
@@ -475,7 +483,8 @@ impl DeltaLayer {
     /// Open the underlying file and read the metadata into memory, if it's
     /// not loaded already.
     ///
-    fn load(&self) -> Result<RwLockReadGuard<DeltaLayerInner>> {
+    fn load(&self, ctx: &RequestContext) -> Result<RwLockReadGuard<DeltaLayerInner>> {
+        self.access_stats.record_access(todo!(), ctx);
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/delta_layer.rs:487]: Should probably do this only in the case that this thread loads (review right hand line 500)

Comment (koivunej) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:487]: Nevermind, this is called through all interesting parts (before iters, get_reconstruct_data) and we are not interested in the actual load, but the access. Looking into how it's ensured that it happens before iters, but resolving for now.

**Comment by problame** [comment]
force-push : squashed & rebased

**Comment by problame** [comment]
force-push: factored out RequestContext plumbing into https://github.com/neondatabase/neon/pull/3504

**Comment by problame** [comment]
I'm finished with the extractions.
Did one pass of self-review.

Let's do some perf analysis before we polish this up.

File: pageserver/src/tenant/storage_layer.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs]: `Instant` has no meaning outside of the process in which it was generated (`CLOCK_MONOTONIC`).
We want to communicate this timestamp to the outside world, so, a wall clock timestamp is needed.

That being said, we could store `Instant`s internally and change the public  API to return

```
age: Duration, // Instant::now() - first_access 
not_accessed_for: Duration // Instant::how() - last_access
```

I think for a dump endpoint, absolute timestamps are better, though.

For a human playing with this, relative times as proposed above would actually be more intuitive.

File: pageserver/src/tenant/storage_layer.rs [hunk 3]
```diff
@@ -188,7 +292,7 @@ pub trait PersistentLayer: Layer {
     /// current_physical_size is computed as the som of this value.
     fn file_size(&self) -> Option<u64>;
 
-    fn info(&self) -> HistoricLayerInfo;
+    fn info(&self, reset: Option<LayerAccessStatsReset>) -> HistoricLayerInfo;
```
Comment (SomeoneToIgnore) [inline pageserver/src/tenant/storage_layer.rs:295]: Some comment on how that's used would be good to have.

File: pageserver/src/tenant/storage_layer.rs [hunk 4]
```diff
@@ -81,6 +88,103 @@ pub enum ValueReconstructResult {
...
+            access_kind: *access_kind,
+        }
+    }
+}
+
+pub static LAYER_ACCESS_STATS_KILLSWITCH: AtomicBool = AtomicBool::new(false);
+
+impl LayerAccessStats {
+    fn record_access(&self, access_kind: LayerAccessKind, task_kind: TaskKind) {
+        if LAYER_ACCESS_STATS_KILLSWITCH.load(atomic::Ordering::SeqCst) {
```
Comment (SomeoneToIgnore) [inline pageserver/src/tenant/storage_layer.rs:137]: What's the idea behind the killswitch here only?
Other methods are still accessing the lock without the switch, should they consider it too?

**Review by SomeoneToIgnore — APPROVED** [review]
I think it's also important to track how many remote layer accesses are made, that led to downloads.
The idea is to be able to spot layers, that are in a bad loop, being always evicted and then redownloaded. Ideally, the time between these two events could be also tracked.

LGTM otherwise.

File: pageserver/src/tenant/storage_layer.rs [hunk 5]
```diff
@@ -81,6 +88,103 @@ pub enum ValueReconstructResult {
     Missing,
 }
 
+#[derive(Default)]
+pub struct LayerAccessStats(RwLock<LayerAccessStatsInner>);
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer.rs:92]: As discussed, just use `Mutex` here.

Rationale:
- Write is used for competing accesses
- Read blocks writes, used for reading it

Outcome: We don't get any benefit from using a mutex as all access is blocking every other kind.

File: pageserver/src/tenant/storage_layer.rs [hunk 6]
```diff
@@ -81,6 +88,103 @@ pub enum ValueReconstructResult {
...
+            task_kind,
+            access_kind,
+        };
+        inner
+            .first_access
+            .get_or_insert_with(|| this_access.clone());
+        inner.count_by_access_kind[access_kind] += 1;
+        inner.task_kind_flag |= task_kind;
+        inner.last_accesses.push_front(this_access);
+        inner.last_accesses.truncate(10); // TODO make runtime-configurable
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer.rs:152]:
```suggestion
        while inner.last_accesses.len() >= LIMIT {
            inner.last_accesses.pop_back();
        }
        inner.last_accesses.push_front(this_access);
```

Though I guess with limit == 10 this doesn't matter, but in general, push could reallocate, so the limit needs to be checked beforehand. In practice, as limit == 10 is not a power of 2, the capacity of the vecdeque will be 16, though this could be relaxed in future std versions.

Would also prefer to see this type have a custom `impl Default` which would specify `VecDeque::with_capacity(LIMIT)`.

File: pageserver/src/tenant/storage_layer.rs (hunk 1 above)
Comment (koivunej) [reply pageserver/src/tenant/storage_layer.rs]: Aah yes, the conversions I was expecting to find were not there. Resolving.

File: pageserver/src/tenant/storage_layer.rs [hunk 7]
```diff
@@ -80,6 +88,103 @@ pub enum ValueReconstructResult {
...
+}
+
+pub static LAYER_ACCESS_STATS_KILLSWITCH: AtomicBool = AtomicBool::new(false);
+
+impl LayerAccessStats {
+    fn record_access(&self, access_kind: LayerAccessKind, task_kind: TaskKind) {
+        if LAYER_ACCESS_STATS_KILLSWITCH.load(atomic::Ordering::SeqCst) {
+            return;
+        }
+        let mut inner = self.0.write().unwrap();
```
Comment (hlinnaka) [inline pageserver/src/tenant/storage_layer.rs:140]: If you're worried about lock contention, could use `try_write` here and just drop the data point if the lock is not free. The stats don't need to be 100% accurate.

File: pageserver/src/tenant/storage_layer.rs (hunk 3 above)
Comment (hlinnaka) [reply pageserver/src/tenant/storage_layer.rs:295]: For clarity, I'd suggest splitting this into two functions:

```
/// read access stats
fn access_stats(&self) -> HistoricLayerInfo

/// Atomically read the current access stats, and reset them 
fn reset_access_stats(&self, opts: LayerAccessStatsReset) -> HistoricLayerInfo
```

Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:295]: > For clarity, I'd suggest splitting this into two functions:

Was hesitant because it's another method on `PersistentLayer` that needs to be implemented everywhere.

How about instead of `Option<LayerAccessStatsReset>`, it's just `LayerAccessStatsReset::{NoReset,JustTaskKindFlags,AllStats}`. 

That way, the call site is self-documenting. Instead of
```
.access_stats(None)
```
we'd have
```
.access_stats(LayerAccessStatsReset::NoReset)
```

File: pageserver/src/tenant/storage_layer.rs (hunk 4 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:137]: This function is the only one called from the data path.
The others need not be considered since they are only called from the new HTTP endpoints.

(I added this killswitch from the perspective of the potential perf impact of this PR, not "could it crash".)

**Comment by problame** [comment]
> Perhaps only push to 'last_accesses' if the previous access was more than N seconds ago, or something like that. Did you have a specific experiment or analysis in mind that you wanted to do with that information?

The scenario I had in mind was: if there are only background loops that access the layer, those accesses would be relatively infrequent, and the `last_accesses` list would give us a straight-forward answer of "what background loop accessed the layer when". We could then easily correlate the timestamp with the logs.

> I was surprised that you bothered with a kill-switch for this. Are you worried about the overhead of tracking those stats? > Updating a few fields seems innocent enough, but the stats here are actually pretty complicated, with a VecDeque and stuff, so I guess there is some risk there.
> Does the kill switch work? I don't see any code that sets LAYER_ACCESS_STATS_KILLSWITCH.

Yeah exactly, the killswitch is for the (unlikely) case that the runtime overhead of the stats is unbearable.

I did manual testing on my dev machine. I think that's sufficient.

> How much of the overhead can the kill-switch remove? It causes LayerAccessStats::record_access to return quickly, but you still have much of the memory overhead of the LayerAccessStats struct. It's smaller if the VecDeque is empty, but still.

Not much. I wanted to run some benchmarks and use the kilswitch to compare before & after. But the pageserver pagecache got in the way, i.e., not many layer accesses happened.
I'll disable / downsize the pagecache and report results shortly.

> My gut feeling would be to start with simpler stats, maybe just a single last_access field, and don't bother with a kill-switch. That's enough to implement an LRU eviction policy, so we could use that to simulate what would happen with LRU evictiion. I have no objections to collecting more stats, but maybe start small, and add more stats when we have a little more information and know better what we'd like to see.

I'd say if the overhead is not too high, let's throw in stats that seem useful today, then dial it back once we've understood how prod actually behaves. I know there's nothing so permanent as a temporary solution, but I think in this case it is really easy to rip out.

**Comment by problame** [comment]
Did some ad-hoc benchmarking.
tl;dr: no noticable difference, with and without killswitch.
So, should I take it out?

---

Setup:

```
/pg_install/v14/bin/pgbench -i -s 50 "port=55432 user=cloud_admin dbname=postgres"
```

`pageserver.toml`:

```
page_cache_size = 20                
layer_access_stats_disable = $killswitch
```

`pgbench` command: `./pg_install/v14/bin/pgbench -S -c 12 -T 100 -P 1  "port=55432 user=cloud_admin dbname=postgres"`

Scrape layer map once per second using:

```
watch -n 1 "curl 'localhost:9898/v1/tenant/3ff96c2a04c3490285cba2019e69fb51/timeline/604094d9de4bda14dfc8da3c1a73e0e4/layer' | jq '.historic_layers | .[] | .access_stats.most_recent[0].when_millis_since_epoch '  | sort"
```

### Run 1: `$killswitch = false`

```
transaction type: <builtin: select only>              
scaling factor: 50                                    
query mode: simple                                    
number of clients: 12                                 
number of threads: 1                                  
duration: 100 s                                       
number of transactions actually processed: 900728     
latency average = 1.321 ms                            
latency stddev = 0.353 ms                             
initial connection time = 23.509 ms                   
tps = 9009.265822 (without initial connection time)   
```

### Run 2: `$killswitch = true`

```
transaction type: <builtin: select only>                  
scaling factor: 50                                        
query mode: simple                                        
number of clients: 12                                     
number of threads: 1                                      
duration: 100 s                                           
number of transactions actually processed: 892262         
latency average = 1.333 ms                                
latency stddev = 0.367 ms                                 
initial connection time = 26.332 ms                       
tps = 8924.787206 (without initial connection time)       
```

**Comment by hlinnaka** [comment]
>I'd say if the overhead is not too high, let's throw in stats that seem useful today, then dial it back once we've understood how prod actually behaves. I know there's nothing so permanent as a temporary solution, but I think in this case it is really easy to rip out.

Ok, works for me

**Comment by problame** [comment]
- [ ] For scraping, it's helpful to have a reliable means to detect restarts between scrape points. Simple solution: include the pageserver's startup timestamp in the `LayerMapInfo` / `LayerAccessStats` API response

**Comment by problame** [comment]
Here's the change to preserve the access stats across evict & on-demand download: https://github.com/neondatabase/neon/pull/3516

Please review and comment there.

File: pageserver/src/tenant/storage_layer.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/tenant/storage_layer.rs:310]: Yeah the 3-state is fine, agreed on less assorted trait methods the better.

File: pageserver/src/tenant/storage_layer.rs (hunk 7 above)
Comment (koivunej) [reply pageserver/src/tenant/storage_layer.rs]: (Now a Mutex instead of RwLock.)

I think here it doesn't really matter or if it would, contention might be really bad sampling strategy. Read currently is actually more costly than this update, there are no allocations etc. cannot really see how this could create a deadlock either.

File: pageserver/src/tenant/storage_layer.rs (hunk 4 above)
Comment (koivunej) [reply pageserver/src/tenant/storage_layer.rs:149]: The SeqCst global order is giving me headache.. Why is it necessary here instead of Relaxed?

File: pageserver/src/tenant/storage_layer.rs (hunk 7 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs]: Yeah, I'm not worried about lock contention, and dropping stats leaves a margin of uncertainty that I want to avoid.

File: pageserver/src/tenant/storage_layer.rs (hunk 4 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:236]: Yeah, due to the subsequent `.lock()`, `Relaxed` would be fine.

File: pageserver/src/tenant/storage_layer.rs (hunk 3 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs]: Pushed the 3-sate

**Comment by problame** [comment]
>     For scraping, it's helpful to have a reliable means to detect restarts between scrape points. Simple solution: include the pageserver's startup timestamp in the LayerMapInfo / LayerAccessStats API response

Created https://github.com/neondatabase/neon/pull/3527, please comment there.

File: pageserver/src/tenant/storage_layer.rs (hunk 4 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:237]: Decided to remove the killswitch, will mark this as resolved.

**Comment by problame** [comment]
Force-push: rebased & squashed commits & cleaned up commit message

**Review by SomeoneToIgnore — APPROVED** [review]
I somehow feel (mostly due to odd naming and repetition in trait impls I've noted before this PR and tried to replace with generics) that the PR could be made simpler.

Yet, I have no particular issues in mind, after all notes are addressed it looks more organic.
Have pushed a few style nits.

File: pageserver/src/tenant/storage_layer.rs [hunk 8]
```diff
@@ -83,6 +89,201 @@ pub enum ValueReconstructResult {
...
+            access_kind: *access_kind,
+        }
+    }
+}
+
+impl LayerResidenceStatus {
+    /// Residence status for a layer file that only exists on the remote.
+    pub fn evicted() -> Self {
+        LayerResidenceStatus::Evicted {
+            timestamp: SystemTime::now(),
```
Comment (SomeoneToIgnore) [inline pageserver/src/tenant/storage_layer.rs:157]: Curious, why not Utc::now for these cases?

File: pageserver/src/tenant/storage_layer.rs [hunk 9]
```diff
@@ -83,6 +89,201 @@ pub enum ValueReconstructResult {
...
+#[derive(Debug)]
+pub struct LayerAccessStats(Mutex<LayerAccessStatsInner>);
+
+#[derive(Debug, Default, Clone)]
+struct LayerAccessStatsInner {
+    first_access: Option<LayerAccessStatFullDetails>,
+    count_by_access_kind: EnumMap<LayerAccessKind, u64>,
+    task_kind_flag: EnumSet<TaskKind>,
+    last_accesses: HistoryBuffer<LayerAccessStatFullDetails, 16>,
+    last_residence_changes: HistoryBuffer<LayerResidenceStatus, 16>,
```
Comment (problame) [inline pageserver/src/tenant/storage_layer.rs:101]: For each HistoryBuffer, have a corresponding counter of how many events fell off the end of the buffer.
E.g., when inserting the 17th `last_residence_change`, one falls off at the end. Capture that in a counter, so post-processing at least knows that events are missing.

**Review by problame — COMMENTED** [review]
Played aorund with a scrapted set of the stats data, see https://github.com/neondatabase/neon/pull/3540#issuecomment-1416860327 and subsequent comments.

Found the data generally useful; some minor problems noted in this review;

File: pageserver/src/tenant/storage_layer.rs [hunk 10]
```diff
@@ -83,6 +89,201 @@ pub enum ValueReconstructResult {
...
+    task_kind: TaskKind,
+    access_kind: LayerAccessKind,
+}
+
+#[derive(Debug, Clone)]
+pub enum LayerResidenceStatus {
+    Resident {
+        timestamp: SystemTime,
+        /// If `true`, then this resident status marks the birth time of the layer.
+        created: bool,
```
Comment (problame) [inline pageserver/src/tenant/storage_layer.rs:116]: consider having a `loaded` flag, for both `Resident` and `Evicted`. Or better even, an enum

```
struct LayerResidenceEvent {
  status: Resident | Evicted
  reason: Load | Created | Change 
}
```

Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:116]: See https://github.com/neondatabase/neon/pull/3540#issuecomment-1416864942 on rationale for this.

**Review by koivunej — APPROVED** [review]

File: pageserver/src/tenant/storage_layer.rs (hunk 8 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:157]: It doesn't make a difference. `Utc::now()` does `SystemTime::now()` internally ([source](https://docs.rs/chrono/latest/src/chrono/offset/utc.rs.html#67-73)).
And since we repr it as a unix timestamp, it's UTC anyways.

File: libs/pageserver_api/src/models.rs [hunk 11]
```diff
@@ -233,6 +238,82 @@ pub struct LayerMapInfo {
...
+    #[serde(rename = "timestamp_millis_since_epoch")]
+    #[serde_as(as = "serde_with::TimestampMilliSeconds")]
+    timestamp: SystemTime,
+    /// The new residence status of the layer.
+    status: LayerResidenceStatus,
+    /// The reason why we had to record this event.
+    reason: LayerResidenceEventReason,
+}
+
+/// The reason for recording a givwen [`ResidenceEvent`].
```
Comment (SomeoneToIgnore) [inline libs/pageserver_api/src/models.rs:274]:
```suggestion
/// The reason for recording a given [`ResidenceEvent`].
```

**Review by SomeoneToIgnore — APPROVED** [review]
Way more readable data representation, thanks.

**Comment by problame** [comment]
Force-push: addressed @SomeoneToIgnore 's review comments & squashed commit.

File: pageserver/src/tenant/storage_layer.rs [hunk 12]
```diff
@@ -83,6 +91,149 @@ pub enum ValueReconstructResult {
...
+    AllStats,
+}
+
+fn system_time_to_millis_since_epoch(ts: &SystemTime) -> u64 {
+    ts.duration_since(UNIX_EPOCH)
+        .expect("better to die in this unlikely case than report false stats")
+        .as_millis()
+        .try_into()
+        .expect("64 bits is enough for few more years")
+}
```
Comment (shanyp) [inline pageserver/src/tenant/storage_layer.rs:126]: shouldn't this be in some "common" crate of neon ?

**Review by shanyp — APPROVED** [review]

File: pageserver/src/tenant/storage_layer.rs (hunk 12 above)
Comment (problame) [reply pageserver/src/tenant/storage_layer.rs:126]: Resolving this conversation to get this PR merged. We can centralize the various `UNIX_EPOCH` usages in the future through an `Ext` trait in the `libs` crate.
